src/expr/src/relation/join_input_mapper.rs

```diff
@@ -354,6 +354,93 @@ impl JoinInputMapper {
         }
         None
     }
+
+    /// Try to find a consequence `c` of the given expression `e` for the given input.
+    ///
+    /// If we return `Some(c)`, that means
+    ///   1. `c` uses only columns from the given input;
+    ///   2. if `c` doesn't hold on a row of the input then `e` also wouldn't hold;
+    ///   3. if `c` holds on a row of the input then `e` might or might not hold.
+    /// 1. and 2. means that if we have a join with predicate `e` then we can use `c` for
+    /// pre-filtering a join input before the join. However, 3. means that `e` shouldn't be deleted
+    /// from the join predicates, i.e., we can't do a "traditional" predicate pushdown.
+    ///
+    /// Note that "`c` is a consequence of `e`" is the same thing as 2., see
+    /// <https://en.wikipedia.org/wiki/Contraposition>
+    ///
+    /// Example: For
+    /// `(t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6)`
+    /// we find
+    /// `t1.f2 = 3 OR t1.f2 = 5` for t1, and
+    /// `t2.f2 = 4 OR t2.f2 = 6` for t2.
+    ///
+    /// Further examples are TPC-H Q07, Q19, and chbench Q07, Q19.
+    pub fn consequence_for_input(
+        &self,
+        expr: MirScalarExpr,
+        index: usize,
+        equivalences: &[Vec<MirScalarExpr>],
+    ) -> Option<MirScalarExpr> {
+        if let Some(localized) =
+            self.try_map_to_input_with_bound_expr(expr.clone(), index, equivalences)
+        {
+            return Some(localized);
+        } else {
+            match expr {
+                MirScalarExpr::CallVariadic {
+                    func: VariadicFunc::Or,
+                    exprs: or_args,
+                } => {
+                    // Each OR arg should provide a consequence. We collect them, and OR them.
+                    let consequences_per_arg = or_args
+                        .into_iter()
+                        .map(|or_arg| {
+                            mz_ore::stack::maybe_grow(|| {
+                                self.consequence_for_input(or_arg, index, equivalences)
+                            })
+                        })
+                        .collect_vec();
+                    if consequences_per_arg.iter().any(|c| matches!(c, None)) {
+                        return None;
+                    } else {
+                        return Some(MirScalarExpr::CallVariadic {
+                            func: VariadicFunc::Or,
+                            exprs: consequences_per_arg
+                                .into_iter()
+                                .map(|c| c.unwrap())
+                                .collect_vec(),
+                        });
+                    }
```

Comment:
```suggestion
                        .collect::<Option<Vec<_>>>()?;
                    Some(MirScalarExpr::CallVariadic {
                        func: VariadicFunc::And,
                        exprs: consequences_per_arg,
                    })
```

Reply:
Can do the same trick above! `collect::<Option<Vec<_>>>()` will return `None` if any element is `None` and otherwise `Some(Vec))`. This also saves an intermediate allocation.

Reply:
That's a nice trick, but I'm not sure whether it would make this function more readable. The problem is that we can do this trick only in the `Or` branch of the top-level match, but not the `And` branch (because that uses `.all` instead of `.any`). If we do it only in the `Or` branch, that would make the `Or` and `And` branches differ more, which would obscure the correspondence between them.

Reply:
Oh I see. I think for the `all` case, you could filter and collect and check is the result is non-empty. I think avoiding intermediate allocations is important, but I don't want to be prescriptive!

Reply:
Ok, I changed both branches.